src/ast/scopes/ParameterScope.ts

```diff
@@ -1,22 +1,41 @@
 import Identifier from '../nodes/Identifier';
 import { EntityPathTracker } from '../utils/EntityPathTracker';
-import ParameterVariable from '../variables/ParameterVariable';
+import { UNKNOWN_EXPRESSION } from '../values';
+import LocalVariable from '../variables/LocalVariable';
 import Scope from './Scope';
 
 export default class ParameterScope extends Scope {
 	parent: Scope;
+	hoistedBodyVarScope: Scope;
 
-	private parameters: ParameterVariable[] = [];
+	private parameters: LocalVariable[] = [];
+	private deoptimizationTracker: EntityPathTracker;
+
+	constructor(parent: Scope, deoptimizationTracker: EntityPathTracker) {
+		super(parent);
+		this.deoptimizationTracker = deoptimizationTracker;
+		this.hoistedBodyVarScope = new Scope(this);
+	}
 
 	/**
 	 * Adds a parameter to this scope. Parameters must be added in the correct
 	 * order, e.g. from left to right.
-	 * @param {Identifier} identifier
-	 * @returns {Variable}
 	 */
-	addParameterDeclaration(identifier: Identifier, deoptimizationTracker: EntityPathTracker) {
-		const variable = new ParameterVariable(identifier, deoptimizationTracker);
-		this.variables[identifier.name] = variable;
+	addParameterDeclaration(identifier: Identifier) {
+		const name = identifier.name;
+		let variable;
+		if (name in this.hoistedBodyVarScope.variables) {
+			variable = this.hoistedBodyVarScope.variables[name] as LocalVariable;
+			variable.addDeclaration(identifier, null);
+		} else {
+			variable = new LocalVariable(
+				name,
+				identifier,
+				UNKNOWN_EXPRESSION,
+				this.deoptimizationTracker
+			);
+		}
+		this.variables[name] = variable;
```

Comment:
Actually not. The difference with keeping this line outside is that even if the parameter has a declaration in the hoisted body scope, it still needs to be visible in the parameter scope because *another parameter may use it as a default*.

I added a test `deconflict-parameter-defaults` that includes all the combinatorics around this and actually turns red if this line is moved into the else block. If the line moved then in this test, the default is wrongly associated with an outside variable of the same name and hence renamed together with this variable.

Reply:
Nice catch ;-)